### PR TITLE
[jsk_perception] add FCN-based classifiers

### PR DESCRIPTION
**Pull request by knorth55**
## What is this?

Classify object based on FCN object segmentation results.

- `label_image_classifier.py`
  - Classify with label image from FCN

- `probability_image_classifier.py`
  - Classify with probability image from FCN

## Sample

```
roslaunch jsk_perception sample_fcn_object_segmentation.launch
```

### `label_image_classifier.py` Result

label_image_classifier

```
$ rostopic echo /label_image_classifier/output -n1
header:
  seq: 45
  stamp:
    secs: 1499198874
    nsecs: 157330989
  frame_id: camera
labels: [7]
label_names: ['car']
label_proba: [0.37124910950660706]
probabilities: [0.0, 0.0, 0.09549389034509659, 0.0, 0.0, 0.003839288605377078, 0.0, 0.37124910950660706, 0.0, 0.0, 0.0, 0.0, 0.34513184428215027, 0.0, 0.0, 0.18428586423397064]
classifier: label_image_classifier
target_names: ['background', 'aeroplane', 'bicycle', 'bird', 'boat', 'bottle', 'bus', 'car', 'cat', 'chair', 'cow', 'diningtable', 'dog', 'horse', 'motorbike', 'person', 'potted plant', 'sheep', 'sofa', 'train', 'tv/monitor']
---
```

### `probability_image_classifier.py` Result

probability_image_classifier

```
$ rostopic echo /probability_image_classifier/output -n1
header:
  seq: 52
  stamp:
    secs: 1499198737
    nsecs: 559743881
  frame_id: camera
labels: [7]
label_names: ['car']
label_proba: [0.36887428164482117]
probabilities: [0.0, 0.0005945690791122615, 0.09234008938074112, 0.0002513318613637239, 0.00020558532560244203, 0.011201946996152401, 0.00019399565644562244, 0.36887428164482117, 0.00016630797472316772, 0.001879233866930008, 0.0009322772966697812, 0.0011920616962015629, 0.33322373032569885, 0.002972929272800684, 0.0003651165752671659, 0.18175159394741058, 0.0006576119340024889, 0.0005574731039814651, 0.0023450853768736124, 0.00019342615269124508, 0.00010132216266356409]
classifier: probability_image_classifier
target_names: ['background', 'aeroplane', 'bicycle', 'bird', 'boat', 'bottle', 'bus', 'car', 'cat', 'chair', 'cow', 'diningtable', 'dog', 'horse', 'motorbike', 'person', 'potted plant', 'sheep', 'sofa', 'train', 'tv/monitor']
---
```

### CODE REVIEW

File: jsk_perception/node_scripts/probability_image_classifier.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    def __init__(self):
+        super(ProbabilityImageClassifier, self).__init__()
+
+    def _classify(self, proba_img):
+        class_proba = proba_img.sum(axis=(0, 1)).astype(np.float32)
+        class_proba[self.ignore_labels] = 0
+        label = np.argmax(class_proba)
+        label_proba = class_proba.max() / class_proba.sum()
+        return label, label_proba
+
```
Comment (wkentaro) [inline jsk_perception/node_scripts/probability_image_classifier.py]: Please add 2 new lines.

File: jsk_perception/sample/sample_fcn_object_segmentation.launch [hunk 2]
```diff
@@ -16,6 +16,24 @@
     <arg name="GPU" value="$(arg GPU)" />
```
Comment (wkentaro) [inline jsk_perception/sample/sample_fcn_object_segmentation.launch]: I think it is better if you put separated launch file named `sample_<node_name>.launch`. What do you think?

File: doc/jsk_perception/nodes/probability_classifier.rst [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+  Ignoring labels.
+
+
+Example
+-------
+
+.. code-block:: bash
+
+   roslaunch jsk_perception sample_alexnet_object_recognition.launch
```
Comment (wkentaro) [inline doc/jsk_perception/nodes/probability_classifier.rst]: Please fix this.

File: doc/jsk_perception/nodes/probability_classifier.rst [hunk 4]
```diff
@@ -0,0 +1,40 @@
+probability_image_classifier.py
+===============================
+
+
```
Comment (wkentaro) [inline doc/jsk_perception/nodes/probability_classifier.rst]: Please add visualized image.
Maybe you can use `draw_classification_result.py`.

File: jsk_perception/sample/sample_fcn_object_segmentation.launch (hunk 2 above)
Comment (knorth55) [inline jsk_perception/sample/sample_fcn_object_segmentation.launch]: ok

**Comment by knorth55** [comment]
Updated and reflect all reviews

File: doc/jsk_perception/nodes/label_image_classifier.rst [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+
+What is this?
+-------------
+
+.. image:: ./images/label_image_classifier.png
+
+Classify from label image.
+
+Return largest segmented label.
```
Comment (k-okada) [inline doc/jsk_perception/nodes/label_image_classifier.rst]: what is the definition of 'largest' ? number of pixels? number of pixels times probability?

File: jsk_perception/sample/sample_probability_image_classifier.launch [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  </node>
+
+  <include file="$(find jsk_perception)/launch/fcn_object_segmentation.launch">
+    <arg name="INPUT_IMAGE" value="image_publisher/output" />
+    <arg name="GPU" value="$(arg GPU)" />
+  </include>
+
+  <node name="probability_image_classifier"
+      pkg="jsk_perception" type="probability_image_classifier.py">
+    <remap from="~input" to="fcn_object_segmentation/output/proba_image" />
```
Comment (wkentaro) [inline jsk_perception/sample/sample_probability_image_classifier.launch]: Why not include sample_fcn_object_segmentation.launch?

File: doc/jsk_perception/nodes/label_image_classifier.rst (hunk 5 above)
Comment (knorth55) [inline doc/jsk_perception/nodes/label_image_classifier.rst]: number of pixel in this case.

File: jsk_perception/sample/sample_probability_image_classifier.launch (hunk 6 above)
Comment (knorth55) [inline jsk_perception/sample/sample_probability_image_classifier.launch]: ?
I include it at the top of this launch.

Comment (wkentaro) [inline jsk_perception/sample/sample_probability_image_classifier.launch]: Ah, I misunderstood by the listed `~target_names`.
Then you can do `<remap from="~target_names" to="fcn_object_segmentation/label_names" />`.

Comment (knorth55) [inline jsk_perception/sample/sample_probability_image_classifier.launch]: ok, I updated

File: jsk_perception/node_scripts/probability_image_classifier.py (hunk 1 above)
Comment (knorth55) [inline jsk_perception/node_scripts/probability_image_classifier.py]: updated

File: jsk_perception/sample/sample_probability_image_classifier.launch (hunk 6 above)
Comment (wkentaro) [inline jsk_perception/sample/sample_probability_image_classifier.launch]: It seems that there is still `~target_names` in <rosparam>

Comment (knorth55) [inline jsk_perception/sample/sample_probability_image_classifier.launch]: Updated

**Comment by knorth55** [comment]
I enable FCN test with `GPU=-1` and all tests has passed!

File: jsk_perception/sample/sample_label_image_classifier.launch [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    <arg name="GPU" value="$(arg GPU)" />
+  </include>
+
+  <node name="label_image_classifier"
+      pkg="jsk_perception" type="label_image_classifier.py">
+    <remap from="~input" to="fcn_object_segmentation/output" />
+    <rosparam>
+      ignore_labels:
+        - 0
+      target_names:
```
Comment (wkentaro) [inline jsk_perception/sample/sample_label_image_classifier.launch]: Please make a yaml file or do the remapping

File: jsk_perception/sample/sample_probability_image_classifier.launch [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    <rosparam>
+      ignore_labels:
+        - 0
+      target_names:
```
Comment (wkentaro) [inline jsk_perception/sample/sample_probability_image_classifier.launch]: same here

**Comment by knorth55** [comment]
@wkentaro can you review again?

**Comment by knorth55** [comment]
@k-okada could you please merge this?

**Comment by knorth55** [comment]
rebased to resolve conflicts and updated to pass test

**Pull request merged**